core/xmodel/cross_query_cache.go

```diff
@@ -0,0 +1,269 @@
+	ctx, _ := context.WithTimeout(context.TODO(), timeOut*time.Second)
+	conn, err := NewEndorsorConn(ce.GetHost())
+	if err != nil {
+		cqc.lg.Info("endorsorQuery NewEndorsorClient error", "err", err.Error())
+		return nil, err
+	}
+	defer conn.Close()
+	cli := pb.NewXendorserClient(conn)
+	endorsorRes, err := cli.EndorserCall(ctx, req)
+	if err != nil {
+		cqc.lg.Info("endorsorQuery EndorserCall error", "err", err.Error())
+		return nil, err
+	}
+	res := &pb.CrossQueryResponse{}
+	err = json.Unmarshal(endorsorRes.GetResponseData(), res)
+	if err != nil {
+		cqc.lg.Info("endorsorQuery Unmarshal error", "err", err)
+		return nil, err
+	}
+	queryRes := &queryRes{
+		res,
+		endorsorRes.GetEndorserSign(),
+	}
+	return queryRes, nil
+}
+
+// 验证CrossQuery背书信息
+func (cqc *CrossQueryCache) isCossQueryValid(
+	crossQueryRequest *pb.CrossQueryRequest,
+	queryMeta *pb.CrossQueryMeta,
+	queryInfo *pb.CrossQueryInfo) bool {
+	// check req from bridge and req from preexec
+	if !isMsgEqual(crossQueryRequest, queryInfo.GetRequest()) {
+		cqc.lg.Info("isCossQueryValid isMsgEqual not equal")
+		return false
+	}
+
+	// check endorsor info
+	signs, ok := cqc.isEndorsorInfoValid(queryMeta, queryInfo.GetSigns())
+	if !ok {
+		cqc.lg.Info("isEndorsorInfoValid not ok")
+		return false
+	}
+
+	// check endorsor sign
+	if !cqc.isEndorsorSignValid(signs, queryInfo) {
+		cqc.lg.Info("isEndorsorSignValid not ok")
+		return false
+	}
+	return true
+}
+
+func (cqc *CrossQueryCache) isEndorsorInfoValid(queryMeta *pb.CrossQueryMeta, signs []*pb.SignatureInfo) ([]*pb.SignatureInfo, bool) {
+	signMap := map[string]*pb.SignatureInfo{}
+	for idx := range signs {
+		signMap[signs[idx].GetPublicKey()] = signs[idx]
+	}
+
+	endorsorMap := map[string]*pb.CrossEndorsor{}
+	endorsors := queryMeta.GetEndorsors()
+	for idx := range endorsors {
+		endorsorMap[endorsors[idx].GetPubKey()] = endorsors[idx]
+	}
+	signsValid := []*pb.SignatureInfo{}
+	for k, v := range signMap {
+		if endorsorMap[k] != nil {
+			signsValid = append(signsValid, v)
+		}
+	}
+	if len(signsValid) < int(queryMeta.GetChainMeta().GetMinEndorsorNum()) {
+		cqc.lg.Info("isEndorsorInfoValid failed")
+		return nil, false
+	}
+	return signsValid, true
+}
+
+func (cqc *CrossQueryCache) isEndorsorSignValid(signsValid []*pb.SignatureInfo, queryInfo *pb.CrossQueryInfo) bool {
+	reqData, err := json.Marshal(queryInfo.GetRequest())
+	if err != nil {
+		cqc.lg.Info("Marshal Request failed", "err", err)
+		return false
+	}
+	resData, err := json.Marshal(queryInfo.GetResponse())
+	if err != nil {
+		cqc.lg.Info("Marshal Response failed", "err", err)
+		return false
+	}
+	cryptoClient, err := crypto_client.CreateCryptoClient(crypto_client.CryptoTypeDefault)
+	data := append(reqData[:], resData[:]...)
+	digest := hash.UsingSha256(data)
+	for idx := range signsValid {
+		pk, err := cryptoClient.GetEcdsaPublicKeyFromJSON([]byte(signsValid[idx].GetPublicKey()))
```

Comment:
Do we checked if the endorser address and public key are mach anywhere?

Reply:
`crossQueryNaming` contract store the  public key, so I can use  public key rather than address to verify directly.